File: app/src/main/java/com/tablet/hid/ui/community/CommunityViewModel.kt
package com.tablet.hid.ui.community

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.tablet.hid.model.CommunityConfigRecord
import com.tablet.hid.model.CommunityUploadBody
import com.tablet.hid.util.ConfigApiClient
import com.tablet.hid.util.CommunityConfigCache
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class CommunityUiState(
    val isLoading: Boolean = false,
    val error: String? = null,
    val configs: List<CommunityConfigRecord> = emptyList(),
    val filterMode: String? = null,
    val filterPlatform: String? = null,
    val filterCategory: String? = null,
    val sortOrder: String = "recent",
    val searchQuery: String = "",
)

class CommunityViewModel(application: Application) : AndroidViewModel(application) {

    private val cache = CommunityConfigCache(application)
    private val _uiState = MutableStateFlow(CommunityUiState())
    val uiState: StateFlow<CommunityUiState> = _uiState.asStateFlow()

    init {
        val cached = cache.getAll()
        if (cached.isNotEmpty()) {
            _uiState.update { it.copy(configs = applyFilters(cached, it)) }
        }
    }

    fun setFilterMode(mode: String?) {
        _uiState.update { state ->
            state.copy(filterMode = mode, configs = applyFilters(cache.getAll(), state.copy(filterMode = mode)))
        }
    }

    fun setFilterPlatform(platform: String?) {
        _uiState.update { state ->
            state.copy(filterPlatform = platform, configs = applyFilters(cache.getAll(), state.copy(filterPlatform = platform)))
        }
    }

    fun setFilterCategory(category: String?) {
        _uiState.update { state ->
            state.copy(filterCategory = category, configs = applyFilters(cache.getAll(), state.copy(filterCategory = category)))
        }
    }

    fun setSortOrder(order: String) {
        _uiState.update { state ->
            state.copy(sortOrder = order, configs = applyFilters(cache.getAll(), state.copy(sortOrder = order)))
        }
    }

    fun setSearchQuery(query: String) {
        _uiState.update { state ->
            state.copy(searchQuery = query, configs = applyFilters(cache.getAll(), state.copy(searchQuery = query)))
        }
    }

    fun refresh() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            val state = _uiState.value
            val result = ConfigApiClient.fetchConfigs(
                mode     = state.filterMode,
                platform = state.filterPlatform,
                category = state.filterCategory,
                sort     = state.sortOrder,
                limit    = 100,
            )
            result.fold(
                onSuccess = { response ->
                    cache.clear()
                    cache.replaceAll(response.configs)
                    response.latestAt?.let { cache.setLatestAt(it) }
                    _uiState.update { s ->
                        s.copy(
                            isLoading = false,
                            error = null,
                            configs = applyFilters(response.configs, s),
                        )
                    }
                },
                onFailure = { e ->
                    _uiState.update { it.copy(isLoading = false, error = e.message) }
                },
            )
        }
    }

    fun syncIfStale() {
        viewModelScope.launch {
            val since = cache.getLatestAt()
            _uiState.update { it.copy(isLoading = true, error = null) }
            val result = ConfigApiClient.fetchConfigs(
                sort  = "recent",
                limit = 100,
                since = since,
            )
            result.fold(
                onSuccess = { response ->
                    cache.mergeIn(response.configs)
                    response.latestAt?.let { cache.setLatestAt(it) }
                    val all = cache.getAll()
                    _uiState.update { s ->
                        s.copy(
                            isLoading = false,
                            error = null,
                            configs = applyFilters(all, s),
                        )
                    }
                },
                onFailure = { e ->
                    _uiState.update { it.copy(isLoading = false, error = e.message) }
                },
            )
        }
    }

    fun uploadConfig(body: CommunityUploadBody): Flow<Result<String>> = flow {
        emit(ConfigApiClient.uploadConfig(body))
    }

    private fun applyFilters(all: List<CommunityConfigRecord>, state: CommunityUiState): List<CommunityConfigRecord> {
        var result = all
        state.filterMode?.let { mode -> result = result.filter { it.mode == mode } }
        state.filterPlatform?.let { platform -> result = result.filter { it.platform == platform } }
        state.filterCategory?.let { category -> result = result.filter { it.category == category } }
        if (state.searchQuery.isNotBlank()) {
            val q = state.searchQuery.lowercase()
            result = result.filter { record ->
                record.profileName.lowercase().contains(q) ||
                record.description?.lowercase()?.contains(q) == true ||
                record.deviceName?.lowercase()?.contains(q) == true ||
                record.tags.any { it.lowercase().contains(q) }
            }
        }
        result = when (state.sortOrder) {
            "popular" -> result.sortedByDescending { it.downloadCount }
            else      -> result.sortedByDescending { it.uploadedAt }
        }
        return result
    }
}

File: ios/TabletHID/ViewModels/CommunityViewModel.swift
import Foundation
import Observation

/// Observable view model for the community config-sharing feature.
/// Mirrors CommunityViewModel.kt: same state shape, same filter/sort logic, same
/// refresh/syncIfStale strategy against the local CommunityConfigCache.
@Observable
@MainActor
final class CommunityViewModel {

    // MARK: - State

    var isLoading: Bool = false
    var error: String?  = nil
    var configs: [CommunityConfigRecord] = []
    var filterMode: String?     = nil
    var filterPlatform: String? = nil
    var filterCategory: String? = nil
    var sortOrder: String       = "recent"
    var searchQuery: String     = ""

    // MARK: - Private

    private let cache = CommunityConfigCache()

    init() {
        let cached = cache.getAll()
        if !cached.isEmpty {
            configs = applyFilters(cached)
        }
    }

    // MARK: - Filter / sort setters

    func setFilterMode(_ mode: String?) {
        filterMode = mode
        configs = applyFilters(cache.getAll())
    }

    func setFilterPlatform(_ platform: String?) {
        filterPlatform = platform
        configs = applyFilters(cache.getAll())
    }

    func setFilterCategory(_ category: String?) {
        filterCategory = category
        configs = applyFilters(cache.getAll())
    }

    func setSortOrder(_ order: String) {
        sortOrder = order
        configs = applyFilters(cache.getAll())
    }

    func setSearchQuery(_ query: String) {
        searchQuery = query
        configs = applyFilters(cache.getAll())
    }

    // MARK: - Network

    /// Full refresh: replaces the cache with a fresh page from the server.
    func refresh() async {
        isLoading = true
        error = nil
        do {
            let response = try await ConfigApiClient.shared.fetchConfigs(
                mode:     filterMode,
                platform: filterPlatform,
                category: filterCategory,
                sort:     sortOrder,
                limit:    100
            )
            cache.clear()
            cache.replaceAll(response.configs)
            if let at = response.latestAt { cache.setLatestAt(at) }
            isLoading = false
            configs = applyFilters(response.configs)
        } catch {
            isLoading = false
            self.error = error.localizedDescription
        }
    }

    /// Incremental sync: fetches only records newer than the stored cursor.
    func syncIfStale() async {
        let since = cache.getLatestAt()
        isLoading = true
        error = nil
        do {
            let response = try await ConfigApiClient.shared.fetchConfigs(
                sort:  "recent",
                limit: 100,
                since: since
            )
            cache.mergeIn(response.configs)
            if let at = response.latestAt { cache.setLatestAt(at) }
            let all = cache.getAll()
            isLoading = false
            configs = applyFilters(all)
        } catch {
            isLoading = false
            self.error = error.localizedDescription
        }
    }

    /// Upload a config to the server. Returns the new record id on success.
    func uploadConfig(_ body: CommunityUploadBody) async throws -> String {
        return try await ConfigApiClient.shared.uploadConfig(body: body)
    }

    /// Fetching by id increments the server-side download count; use this before import.
    func fetchConfigForImport(id: String) async throws -> CommunityConfigRecord {
        let record = try await ConfigApiClient.shared.fetchConfig(id: id)
        cache.replaceOrInsert(record)
        configs = applyFilters(cache.getAll())
        return record
    }

    // MARK: - Filtering (client-side, same logic as Android)

    private func applyFilters(_ all: [CommunityConfigRecord]) -> [CommunityConfigRecord] {
        var result = all
        if let mode = filterMode, !mode.isEmpty {
            result = result.filter { $0.mode == mode }
        }
        if let platform = filterPlatform, !platform.isEmpty {
            result = result.filter { $0.platform == platform }
        }
        if let category = filterCategory, !category.isEmpty {
            result = result.filter { $0.category == category }
        }
        if !searchQuery.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            let q = searchQuery.lowercased()
            result = result.filter { record in
                record.profileName.lowercased().contains(q) ||
                (record.description?.lowercased().contains(q) ?? false) ||
                (record.deviceName?.lowercased().contains(q) ?? false) ||
                record.tags.contains(where: { $0.lowercased().contains(q) })
            }
        }
        switch sortOrder {
        case "popular":
            result.sort { $0.downloadCount > $1.downloadCount }
        default:
            result.sort { $0.uploadedAt > $1.uploadedAt }
        }
        return result
    }
}
